b2/sync.py

```diff
@@ -761,64 +762,123 @@ def check_file_replacement(source_file, dest_file, args):
         raise CommandError('Invalid option for --compareVersions')
 
 
+def should_transfer_file(source_file, dest_file, args):
+    """
+    Decides whether to transfer the file from the source to the destination.
+    """
+    if source_file is None:
+        # No source file.  Nothing to transfer.
+        return False
+    elif dest_file is None:
+        # Source file exists, but no destination file.  Always transfer.
+        return True
+    else:
+        # Both exist.  Transfer only if the two are different.
+        return files_are_different(source_file, dest_file, args)
+
+
+def make_b2_delete_actions(source_file, dest_file, dest_folder, transferred):
+    for (version_index, version) in enumerate(dest_file.versions):
+        keep = (version_index == 0) and (source_file is not None) and not transferred
+        if not keep:
+            note = ''
+            if version.action == 'hide':
+                note = '(hide marker)'
+            elif transferred or 0 < version_index:
+                note = '(old version)'
+            yield B2DeleteAction(
+                dest_file.name, dest_folder.make_full_path(dest_file.name), version.id_, note
+            )
+
+
+def make_b2_keep_days_actions(
+    source_file, dest_file, dest_folder, transferred, keep_days, now_millis
+):
+    """
+    Creates the actions to hide or delete existing versions of a file
+    stored in B2.
+
+    When keepDays is set, all files that were visible any time from
+    keepDays ago until now must be kept.  If versions were uploaded 5
+    days ago, 15 days ago, and 25 days ago, and the keepDays is 10,
+    only the 25-day old version can be deleted.  The 15 day-old version
+    was visible 10 days ago.
+    """
+    prev_age_days = None
+    deleting = False
+    for (version_index, version) in enumerate(dest_file.versions):
+        # How old is this version?
+        age_days = (now_millis - version.mod_time) / ONE_DAY_IN_MS
+
+        # We assume that the versions are ordered by time, newest first.
+        assert prev_age_days is None or prev_age_days <= age_days
+
+        # Do we need to hide this version?
+        if version_index == 0 and source_file is None and version.action == 'upload':
+            yield B2HideAction(dest_file.name, dest_folder.make_full_path(dest_file.name))
+
+        # Can we start deleting?  Once we start deleting, all older
+        # versions will also be deleted.
```

Comment:
This relies on the fact that `dest_file.versions` are sorted by `mod_time` attribute and this is not asserted, nor documented here or in `file_version.py`. We should at least document it thoroughly, because if the ordering changes and this function will not be updated, it can cause data loss. That kind of bug could also be easily missed by a test, which may accidentally keep the ordering after it is changed in the code due to small homogeneous sample test data.